Add infrastructure for proxy templates

andig opened the pull request:
**Preface**

Evcc unterstützt `native` und `custom` Konfigurationen (Configs). `native` Configs sind im evcc core implementiert. `custom` Configs zeichnen sich durch beliebig komplexe Verknüpfung von unterschiedlichen Plugins aus.

**Design goals**

1. Vereinfachung der Konfiguration für den Endanwender
    - `custom` Configs werden durch leichter anwendbare Proxies ersetzt. 
    - Proxies ergänzen evcc um weitere pseudo-native Konfigurationstypen. 
    - Proxies werden durch yaml Templates im config Repo beschrieben. 
    - Die finale `custom` Konfiguration wird aus den Proxies zur Laufzeit erzeugt, Anwender kommen damit nicht in Kontakt.

3. Automatische Fehlerbehebung
    - da der finale `custom` Konfiguration erst zur Laufzeit generiert wird, können Fehlerbehebungen in der Instanziierung der Proxies erfolgen und mit neuer evcc Version ausgeliefert werden.
    
4. Geführte Konfiguration
    - Proxies werden durch `params` ergänzt die alle notwendigen Konfigurationseigenschaften enthalten.
    - Für `native` Configs werden die `params` aus evcc selbst erzeugt -> Vermeidung von Doppelpflege
    
3. Verbesserte Dokumentation
    - händisch gepflegter Rahmen im docs Repo
    - Einbindung der Proxies wie auch `native` Configs mittels Repo Link (@naltatis)
    
4. Vermeidung von Doppelpflege
    - Für `native` Configs werden die `params` aus evcc selbst erzeugt. 
    - Es erfolgt keine(!) Doppelpflege im Config Repo wie derzeit

Die weitere Vorgehensweise könnte jetzt so aussehen (grobe Reihenfolge):

- [x] Entscheidung: proxy Konfigurationen werden in evcc geflegt, config repo wird perspektivisch ausgephast
- [ ] Liste aller relevanten komplexen Configs erstellen /cc @premultiply 
- [ ] alle relevanten configs werden mit Nutzung von `params` und `render` neu in evcc erstellt
- [x] config generator erzeugt Go Code der Zugang zu `params` und `render` gibt
- [ ] cli wird in evcc integriert- evcc stellt dafür einfache und komplexe Konfigurationen bereit
- [x] Erzeugung der proxy Konfigurationstemplates aus evcc mittels Generator
- [ ] Erzeugung der einfachen Konfigurationstemplates aus evcc mittels Generator
- [ ] Import der Templates ins docs repo
- [ ] config repo archivieren

andig committed:
Add draft infrastrucuture for providing complex templates

andig committed:
wip

andig committed:
wip

andig commented on meter/custom/sunspec.yaml:
```diff
@@ -0,0 +1,24 @@
+  uri: {{ .uri }}
+  id: {{ .id }}
+  {{- if eq .usage "grid" }}
+  power: 203:W # sunspec model 203 meter
```
Hier wird auch die Limitierung klar: für den Anwender ist es jetzt nicht mehr einfach möglich, aus 203 213 zu machen falls er nicht `float` nutzt. Lösbar wäre das über einen weiteren Parameter mit Defaultwert. Das Wissen über die Detailkonfiguration bei den Anwendern geht damit am Ende verloren...

DerAndereAndi replied:
Wer auf der Ebene etwas anpassen will/muss, sollte sich dann in der Doku die komplexe Syntax anschauen und diese nehmen. Ich sehe hier eher den Vorteil wenn sich so ein Detail als falsch herausstellt, werden alle mit dem Update ohne Anpassungen den Fix bekommen.

DerAndereAndi commented:
Danke @andig Ich schaue es mir später im Detail an, daher kann ich noch kein inhaltliches Feedback geben.

andig committed:
Update sunspec.yaml

DerAndereAndi committed:
Include any yaml template in the folder

Also added another example

DerAndereAndi commented:
Habe mal etwas hinzugefügt womit ich hier testen kann

andig commented on meter/custom/sunspec.yaml:
```diff
@@ -0,0 +1,24 @@
+- name: uri
+- name: id
+  default: 1
+# sample: |
```
Idee: Der `sample` Block ist egtl. unnötig. Der ergibt sich ja automatisch aus `type` und `params`. Oder haben wir einen Fall wo das nicht ausreichen würde?
Wenn die Annahme stimmt könnte man die Templates aus diesem PR jetzt entfernen und Deinen PR in config sowie den Generator für die README entsprechen umbauen.
`hint` würde ich erstmal grundsätzlich weglassen- Fokus auf Funktion statt UX.

DerAndereAndi replied:
Ne passt. Man kännte auch einfach das weiter `sample` nennen anstatt jetzt bei dir `render`;)

andig commented:
Die weitere Vorgehensweise könnte jetzt so aussehen (grobe Reihenfolge):

- [ ] Entscheidung: config repo wird auf komplexe Konfigurationen beschränkt
- [ ] alle relevanten configs werden auf `params` und `render` umgebaut
- [ ] config generator erzeugt Go Code der Zugang zu `params` und `render` gibt
- [ ] alle komplexen Konfigurationen in evcc importieren (`go get ...`)
- [ ] einfache Konfigurationen werden aus evcc Registry erzeugt - dafür braucht es Zugang zu den Parametern
- [ ] cli wird in evcc integriert- evcc stellt dafür einfache und komplexe Konfigurationen bereit
- [ ] Erzeugung der kompletten Konfigurationstemplates aus evcc (statt aus config)- nur dort sind einfache und komplexe dann verfügbar
- [ ] Import der Templates ins docs repo
- [ ] alle einfachen Konfigurationen aus config repo löschen
- [ ] ggf alle komplexen Konfigurationen aus config nach evcc migrieren und config repo löschen

andig committed:
Add rendering sample from params and remove Params from rendering yaml

andig committed:
remove sample

andig committed:
Improve sample rendering

DerAndereAndi commented on meter/custom/sunspec.yaml:
```diff
@@ -0,0 +1,20 @@
+  power: 203:W # sunspec model 203 meter
+  {{ end -}}
+  {{- if eq .usage "battery" }}
+  soc: ChargeState
```
Diese Notation funktioniert nicht für die Readme, wo wir alle Usages anzeigen wollen.

Ich kreiere für jede angegebe Choice ein `.Usage.grid` bool etc. und frage dann per `{{- if .Usage.grid }}` etc. ab

andig committed:
Render all usage types

andig commented:
**Preface**

Evcc unterstützt `native` und `custom` Konfigurationen (Configs). `native` Configs sind im evcc core implementiert. `custom` Configs zeichnen sich durch beliebig komplexe Verknüpfung von unterschiedlichen Plugins aus.

**Design goals**

1. Vereinfachung der Konfiguration für den Endanwender
    - `custom` Configs werden durch leichter anwendbare Proxies ersetzt. 
    - Proxies ergänzen evcc um weitere pseudo-native Konfigurationstypen. 
    - Proxies werden durch yaml Templates im config Repo beschrieben. 
    - Die finale `custom` Konfiguration wird aus den Proxies zur Laufzeit erzeugt, Anwender kommen damit nicht in Kontakt.

3. Automatische Fehlerbehebung
    - da der finale `custom` Konfiguration erst zur Laufzeit generiert wird, können Fehlerbehebungen in der Instanziierung der Proxies erfolgen und mit neuer evcc Version ausgeliefert werden.
    
4. Geführte Konfiguration
    - Proxies werden durch `params` ergänzt die alle notwendigen Konfigurationseigenschaften enthalten.
    - Für `native` Configs werden die `params` aus evcc selbst erzeugt -> Vermeidung von Doppelpflege
    
3. Verbesserte Dokumentation
    - händisch gepflegter Rahmen im docs Repo
    - Einbindung der Proxies wie auch `native` Configs mittels Repo Link (@naltatis)
    
4. Vermeidung von Doppelpflege
    - Für `native` Configs werden die `params` aus evcc selbst erzeugt. 
    - Es erfolgt keine(!) Doppelpflege im Config Repo wie derzeit

andig committed:
Render usages

andig committed:
wip

andig committed:
Create /templates folder

DerAndereAndi committed:
Fix build

andig commented on templates/template.go:
```diff
@@ -0,0 +1,87 @@
+	_ "embed"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
```
Da fehlt noch

    …/v3

DerAndereAndi replied:
Fixed

DerAndereAndi committed:
Fix sprig version

premultiply commented:
Das wird ein Brett! 💪
Was kann/soll ich tun?

andig commented:
> Das wird ein Brett! 💪
> Was kann/soll ich tun?

Eine Aufgabe wäre mal alle künftig notwendigen "proxy" Templates aufzulisten und ihnen einen Namen zu geben. Jedes dieser Templates müsste dann hier (oder später) aufgenommen werden. Die Namen sind diesmal kritisch da sie sich für den Anwender natürlich nicht mehr ändern dürfen. Sinnvoll ist das nur da, wo die Funktion vollständig in evcc liegt, also keine eingebundenen Skripte etc.

andig committed:
Trim code

DerAndereAndi commented:
Ich baue auch schon ein paar. @premultiply wichtig ist auch dass wir Varianten die heute in mehreren Dateien verteilt sind (grid, pv, battery) in eine Datei zusammenführen können.

naltatis commented:
@premultiply @DerAndereAndi für die neue (noch WIP) Dokuseite habe ich die bestehenden Configs mal in eine Struktur gebracht. Das ist sicher ein cuter Startpunkt für die Benamungen.
https://docs.evcc.io/docs/configuration/meters

Idee war hier eher über den Hersteller (und Produktserie) zu kommen als über die verwendeten Protokolle die ja aktuell noch im Namen der config stehen.

PS: Die Code Samples liegen zwar auch im docs Repo werden dort aber nicht bearbeitet sondern automatisch aus dem config Repo geholt (git subtree).

andig committed:
Add tests

andig committed:
Better test logging

andig commented:
> Idee war hier eher über den Hersteller (und Produktserie) zu kommen als über die verwendeten Protokolle die ja aktuell noch im Namen der config stehen.

Die Protokolle sind perspektivisch auch weniger relevant da der Anwender damit nicht mehr in Berührung kommt- die Proxies werden das weitgehend kapseln. Passt also.

andig committed:
Parallel tests

DerAndereAndi committed:
Add modbus template support

- templates can have the magic comment `# ::modbus-setup::` to define that this is the place for modbus specific parameters
- `params.name: modbus` defines the possible modbus interfaces via `choice: ["rs485", "tcpip"]`
- valid `modbus` values for `evcc.config` are `rs485serial`, `rs485tcpip`, `tcpip`
- Includes more examples for `E3DC`, `Eastron SDM`, `Fronius Gen24 Plus`, `SMA Inverter`, `SMA Meter`, `Solarlog`

andig committed:
Fix tests

andig committed:
add acceptable errors

andig committed:
Add code generator for proxies

andig commented on templates/template.go:
```diff
@@ -0,0 +1,201 @@
+	return nil
+}
+
+func (t Template) contains(slice []string, element string) bool {
```
funk.ContainsString

Siehe https://pkg.go.dev/github.com/thoas/go-funk#ContainsString

DerAndereAndi committed:
Use funk.ContainsString

DerAndereAndi committed:
Partly fix build

It still panics as there are no templates for `charger` and `vehicle` which means those subdirectories under `docs` are missing

andig committed:
Add test property to supplement default

andig committed:
Remove hints and empty default values

andig committed:
Add docs and split modbus handling

andig committed:
Simplify

andig committed:
remove hint

DerAndereAndi committed:
Fix proxy and generate issues with modbus

andig committed:
wip

DerAndereAndi committed:
Fix generate

DerAndereAndi committed:
Update templates

- Add description for usage in CLI configure
- Params have value for usage in CLI configure
- RednerProxyWithValues to render CLI configured templates

DerAndereAndi committed:
Add first implementation of CLI configure task

Invoke via `evcc configure`

DerAndereAndi committed:
Add eebus setup to generated config

andig committed:
Partial merge of config ui code for reuse

andig committed:
wip

andig committed:
Add vehicle config

andig committed:
wip

andig committed:
Export vehicle config

DerAndereAndi committed:
Allow adding multiple battery meters

DerAndereAndi committed:
Automatically assign device names

DerAndereAndi committed:
Actually ask for battery meters

DerAndereAndi committed:
Add template requirements

Allow to define other items to be required for this device to work, e.g. if it is based on EEBUS and thus requires the EEBUS server be setup and running

DerAndereAndi committed:
Merge branch 'master' into demo/params

# Conflicts:
#	vehicle/nissan.go

DerAndereAndi committed:
Fix proxy rendering with templates

DerAndereAndi committed:
Add more working template options

- `mask` for masking user input, e.g. passwords
- `advanced` for defining params the user does not need to be asked (optionals, that have a `default` set) to reduce complexity
- `example` for defining example values, e..g `WVW...` vins, that are not `default` values, because they won't work if they are not provided
- add required check

DerAndereAndi committed:
Remove params from rendering with no values

Remove template params that don't have a default and no value from being rendered

Otherwise optional values that are provided with no values (e.g. vin) would be rendered as `vin:`

DerAndereAndi committed:
Fix batteries list config param name

DerAndereAndi committed:
Update sma template

DerAndereAndi committed:
Merge branch 'master' into demo/params

DerAndereAndi committed:
Add sample charger and vehicle template support

DerAndereAndi committed:
Align template type names

DerAndereAndi committed:
More adjustments

DerAndereAndi committed:
Some more loadpoint additions

- Show charger description and charger config number when describing that the loadpoint for that charger is now to be done
- Show  the vehicle title (or description) when asking if it will charge at this loadpoint
- Added question for the maximum power the charger can provide to derive `mincurrent` and `maxcurrent` from that
- Added question for default charging mode
- Added check if charger provides meter data
- Ask for a separate charge meter if charger provides no meter data

DerAndereAndi committed:
Move charger setup into loadpoint setup

DerAndereAndi committed:
More additions

- Option to validate input data for int values
- Added option to specify max amperage and amount of phases manually

DerAndereAndi committed:
Remove evcc.yaml dependency from configure

DerAndereAndi committed:
Hide proxy debug output

DerAndereAndi committed:
Merge branch 'master' into demo/params

DerAndereAndi commented:
Closing this issue, as we try a slightly different approach.
